src/compiler/checker.ts

```diff
@@ -17594,6 +17600,59 @@ namespace ts {
                 return getTypeWithFacts(mapType(type, narrowTypeForTypeofSwitch(impliedType)), switchFacts);
             }
 
+            function narrowTypeByConstructor(type: Type, constructorAccessExpr: AccessExpression, operator: SyntaxKind, identifier: Expression, assumeTrue: boolean): Type {
+                // Do not narrow when checking inequality.
+                if (!assumeTrue || operator !== SyntaxKind.EqualsEqualsToken && operator !== SyntaxKind.EqualsEqualsEqualsToken) {

```

Comment:
This condition is incorrect. If `!assumeTrue`, we should perform the narrowing only for `!==` and `!=`, while if `assumeTrue`, we should perform narrowing for `===` and `==`. So:

```suggestion
                if (assumeTrue ? (operator !== SyntaxKind.EqualsEqualsToken && operator !== SyntaxKind.EqualsEqualsEqualsToken) : (operator !== SyntaxKind.ExclamationEqualsToken && operator !== SyntaxKind.ExclamationEqualsEqualsToken)) {
```

Reply:
@weswigham I think you're right and I'll make this change as well. 

After looking at this a bit more though, could we be performing narrowing in the false cases too?

For example, given:
```ts
class C1 {}

let var1: C1 | number;
if (var1.constructor == C1) {
    var1; // C1
}
else {
    var1; // number | C1
}

if (var1.constructor != C1) {
    var1; // number | C1
}
else {
    var1; // C1
}
```
In the first `else` body, and the second `if` body, can we remove `C1` from the union since (I think) we know the type can't ever be `C1`?

Reply:
No - given another interface structurally identical to C1, but with a different `constructor` symbol, we'd fall into the false branch, but still "have a C1".